src/main/java/net/minecraftforge/common/brewing/ContainerBrewingRecipe.java

```diff
@@ -0,0 +1,112 @@
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common.brewing;
+
+import com.google.gson.JsonObject;
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.alchemy.PotionUtils;
+import net.minecraft.world.item.crafting.Ingredient;
+import net.minecraft.world.item.crafting.RecipeSerializer;
+import net.minecraft.world.item.crafting.ShapedRecipe;
+import net.minecraftforge.registries.ForgeRegistries;
+import net.minecraftforge.registries.ForgeRegistryEntry;
+import net.minecraftforge.registries.ObjectHolder;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public record ContainerBrewingRecipe(@Nonnull ResourceLocation id,
+                                     @Nonnull Item input,
+                                     @Nonnull Ingredient ingredient,
+                                     @Nonnull Item output) implements IBrewingRecipe
+{
+    @ObjectHolder("forge:container_brewing")
+    public static final RecipeSerializer<ContainerBrewingRecipe> SERIALIZER = null;
+
+    @Override
+    public boolean isInput(@Nonnull ItemStack stack)
+    {
+        return this.input == stack.getItem();
+    }
+
+    @Override
+    public ItemStack getOutput(ItemStack input, ItemStack ingredient)
+    {
+        return isInput(input) && isIngredient(ingredient) ? PotionUtils.setPotion(new ItemStack(output()), PotionUtils.getPotion(input)) : ItemStack.EMPTY;
+    }
+
+    @Override
+    public boolean isIngredient(ItemStack ingredient)
+    {
+        return this.ingredient.test(ingredient);
+    }
+
+    @Override
+    public ItemStack getResultItem()
+    {
+        return new ItemStack(output());
+    }
+
+    @Override
+    public ResourceLocation getId()
+    {
+        return id;
+    }
+
+    @Override
+    public RecipeSerializer<?> getSerializer()
+    {
+        return SERIALIZER;
+    }
+
+    public static class Serializer extends ForgeRegistryEntry<RecipeSerializer<?>> implements RecipeSerializer<ContainerBrewingRecipe>
+    {
+
+        @Override
+        public ContainerBrewingRecipe fromJson(final ResourceLocation p_199425_1_, final JsonObject p_199425_2_)
+        {
+            Item input = ForgeRegistries.ITEMS.getValue(new ResourceLocation(p_199425_2_.get("input_container").getAsString()));
```

Comment:
Declaring the parameter in the constructor as non null does really solve all NPE and Json exceptions here. They should be caught properly and use the `GsonHelper`.